modules/budgets/app/models/budget.rb

```diff
@@ -76,10 +76,19 @@ def initialize(attributes = nil)
 
   def copy_from(arg)
     budget = (arg.is_a?(Budget) ? arg : self.class.find(arg))
-    attrs = budget.attributes.dup
-    super(attrs)
-    self.labor_budget_items = budget.labor_budget_items.map(&:dup)
-    self.material_budget_items = budget.material_budget_items.map(&:dup)
+    self.attributes = budget.attributes.dup
+    self.id = nil if self.id == budget.id
+    self.labor_budget_items = budget.labor_budget_items.map do |lc|
+      lc.dup do |d|
+        d.budget = self && d.budget_id = self.id
+      end
+    end
+    self.material_budget_items = budget.material_budget_items.map do |mc|
+      mc.dup do |d|
+        d.budget = self
+        d.budget_id = self.id
+      end
+    end
```

Comment:
I believe this can be streamlined a bit which should also make the code easier to understand:
```
def self.new_copy(source)
  source = (source.is_a?(Budget) ? source : self.class.find(source))
  copy = new(source.attributes.slice('project_id', 'subject', 'description', 'fixed_date').merge(author => User.current))
  source.labor_budget_items.each do |bi|
    copy.labor_budget_items.build(bi.attributes.slice('units', 'cost_type_id', 'comments', 'amount'))
  end
  source.material_budget_items.each do |bi|
    copy.material_budget_items.build(bi.attributes.slice('units', 'cost_type_id', 'comments', 'amount'))
  end
end
```

There are a couple of changes in here:
1) it is turned into a class method as I believe this method will not be fit to merge data from two budgets but it can create a copy of an existing budget. The call in the controller will have to be adapted.
2) It assigns the current user to be the author of the budget. This I believe to be important as not to suggest that somebody created a budget when it was copied by a different user.